src/wasm2asm.h

```diff
@@ -196,7 +197,116 @@ class Wasm2AsmBuilder {
+  SetLocal* inc =
+      b.makeSetLocal(1, b.makeBinary(addOp,
+                                     b.makeGetLocal(1, argType),
+                                     b.makeConst(Literal(1))));
+  Break* cont = b.makeBreak(loopName);
+  Loop* loop = b.makeLoop(loopName, b.blockify(brIf, update, inc, cont));
+  Block* loopBlock = b.blockifyWithName(loop, blockName);
+  SetLocal* initCount = b.makeSetLocal(1, b.makeConst(Literal(0)));
+  return b.makeFunction(funcName,
+                        std::vector<NameType>{NameType("x", argType)},
+                        argType,
+                        std::vector<NameType>{NameType("count", argType)},
+                        b.blockify(initCount, loopBlock));
+}
+Function* makeRotFunc(MixedArena& allocator, bool isLRot, bool is32Bit=true) {
+  Builder b(allocator);
+  // left rotate is:
+  // (((((~0) >>> k) & x) << k) | ((((~0) << (w - k)) & x) >>> (w - k)))
+  // where k is shift modulo w. reverse shifts for right rotate
+  static Name names[2][2] = {{Name(ROTR64), Name(ROTR32)},
+                             {Name(ROTL64), Name(ROTL32)}};
+  static BinaryOp shifters[2][2] = {{ShrUInt64, ShrUInt32},
+                                    {ShlInt64, ShlInt32}};
+  Name funcName = names[isLRot][is32Bit];
+  BinaryOp lshift = shifters[isLRot][is32Bit];
+  BinaryOp rshift = shifters[!isLRot][is32Bit];
+  BinaryOp orOp = is32Bit ? OrInt32 : OrInt64;
+  BinaryOp andOp = is32Bit ? AndInt32 : AndInt64;
+  BinaryOp subOp = is32Bit ? SubInt32 : SubInt64;
+  WasmType argType = is32Bit ? i32 : i64;
+  auto shiftVal = [&]() {
+    return b.makeBinary(andOp, b.makeGetLocal(1, argType),
+                        b.makeConst(Literal(is32Bit ? (32 - 1) : (64 - 1))));
+  };
+  auto widthSub = [&]() {
+    return b.makeBinary(subOp, b.makeConst(Literal(is32Bit ? 32 : 64)),
+                        shiftVal());
+  };
+  auto fullMask = [&]() {
+    return b.makeConst(is32Bit ? Literal(~(uint32_t)0) : Literal(~(uint64_t)0));
+  };
+
+  Binary* maskRShift = b.makeBinary(rshift, fullMask(), shiftVal());
+  Binary* lowMask = b.makeBinary(andOp, maskRShift, b.makeGetLocal(0, argType));
+  Binary* lowShift = b.makeBinary(lshift, lowMask, shiftVal());
+  Binary* maskLShift = b.makeBinary(lshift, fullMask(), widthSub());
+  Binary* highMask =
+      b.makeBinary(andOp, maskLShift, b.makeGetLocal(0, argType));
+  Binary* highShift = b.makeBinary(rshift, highMask, widthSub());
+  Binary* body = b.makeBinary(orOp, lowShift, highShift);
+  return b.makeFunction(funcName,
+                        std::vector<NameType>{NameType("x", argType),
+                              NameType("k", argType)},
+                        argType,
+                        std::vector<NameType>{},
+                        body);
+}
+
 Ref Wasm2AsmBuilder::processWasm(Module* wasm) {
+  wasm->addFunction(makeCtzFunc(wasm->allocator));
+  wasm->addFunction(makePopcntFunc(wasm->allocator));
+  wasm->addFunction(makeRotFunc(wasm->allocator, true));
```

Comment:
Hm, bool parameters are kinda icky because they tend to not offer context at the call site for what true/false actually means. (https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/06/02/against-bool-parameters/)

How about `makeRotFunc` takes a `WasmType` instead of a bool, we `assert(type == i32 || type == i64)`, and locally in `makeRotFunc` we have `bool is32Bit = type == i32;`

Edit: OH, this true/false is for left/right rotation. Maybe similar advice, but with opcodes instead of types? This actually lets you collapse the two bools, because `RotLInt32` encodes all the information you need.